parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java

```diff
@@ -0,0 +1,1063 @@
+          rc.startGroup();
+          rc.endGroup();
+          rc.endField("missing", 2);
+
+          rc.endGroup();
+          rc.endField("typed_value", 2);
+
+          rc.startField("value", 1);
+          rc.addBinary(Binary.fromConstantByteArray(cValue.getValue()));
+          rc.endField("value", 1);
+
+          rc.startField("metadata", 0);
+          rc.addBinary(expectedMetadata);
+          rc.endField("metadata", 0);
+
+          rc.endGroup();
+          rc.endField("v", 0);
+          rc.endMessage();
+        });
+
+    Schema variantSchema = record(
+        "v",
+        field("metadata", Schema.create(Schema.Type.BYTES)),
+        optionalField("value", Schema.create(Schema.Type.BYTES)));
+    Schema expectedSchema = record("VariantMessage", field("v", variantSchema));
+
+    GenericRecord expectedRecord = instance(
+        expectedSchema,
+        "v",
+        instance(
+            variantSchema,
+            "metadata",
+            expectedMetadata.toByteBuffer(),
+            "value",
+            expectedValue.toByteBuffer()));
+
+    // both should behave the same way
+    assertReaderContains(new AvroParquetReader(new Configuration(), test), expectedSchema, expectedRecord);
+  }
+
+  public <T extends IndexedRecord> void assertReaderContains(
+        AvroParquetReader<T> reader, Schema expectedSchema, T... expectedRecords) throws IOException {
+      for (T expectedRecord : expectedRecords) {
+        T actualRecord = reader.read();
+        Assert.assertEquals("Should match expected schema", expectedSchema, actualRecord.getSchema());
+        Assert.assertEquals("Should match the expected record", expectedRecord, actualRecord);
+      }
+      Assert.assertNull(
+          "Should only contain " + expectedRecords.length + " record" + (expectedRecords.length == 1 ? "" : "s"),
+          reader.read());
+  }
+
+  // We need to store two copies of the schema: one without the Variant type annotation that is used to construct the
+  // Avro schema for writing, and one with type annotation that is used in the actual written parquet schema, and when
+  // reading.
+  private class TestSchema {
+    MessageType parquetSchema;
+    MessageType unannotatedParquetSchema;
+    GroupType variantType;
+    GroupType unannotatedVariantType;
+    TestSchema(Type shreddedType) {
+      variantType = variant("var", 2, shreddedType);
+      unannotatedVariantType = unannotatedVariant("var", 2, shreddedType);
+      parquetSchema = parquetSchema(variantType);
+      unannotatedParquetSchema = parquetSchema(unannotatedVariantType);
+    }
+
+    TestSchema() {
+      variantType = variant("var", 2);
+      unannotatedVariantType = unannotatedVariant("var", 2);
+      parquetSchema = parquetSchema(variantType);
+      unannotatedParquetSchema = parquetSchema(unannotatedVariantType);
+    }
+
+    TestSchema(GroupType variantType, GroupType unannotatedVariantType, MessageType parquetSchema, MessageType unannotatedParquetSchema) {
```

Comment:
Aren't the Parquet schemas just the variant types wrapped using `parquetSchema`? Why pass them in?

Reply:
Nevermind, I see that this is for cases where we are exercising edge cases, like a missing `value` column.